---
title: "Basic RNA-seq analysis (pca and differentially expressed genes)"
author: "Robert Allaway"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    code_fold: hide
    toc: true
    toc_float: true
---

# Setup 
```{r}
library(reticulate)
library(dplyr)
library(readr)
library(tximport)
library(DESeq2)

synapse <- import('synapseclient')
syn <- synapse$Synapse()

syn$login()

cNF_quants <- syn$tableQuery('select * from syn29654184')$filepath %>% read_csv 
pNF_quants <- syn$tableQuery('select * from syn29657306')$filepath %>% read_csv

tx2gene <- syn$get("syn29532380")$path %>% read_tsv(col_names = F)

metadata <- tibble("synid" = c(cNF_quants$id, pNF_quants$id), 
                  "batch" = c(rep("two", nrow(cNF_quants)), rep("one", nrow(pNF_quants))),
                  "type" = c(rep('cNF', 6), rep('pNF', 2), "cNF", "pNF", rep('cNF', 7), rep('pNF', 9)))


files <- sapply(metadata$synid, function(x){
  syn$get(x)$path
})

counts <- tximport::tximport(files = files, 
                             importer = readr::read_tsv,
                             tx2gene = tx2gene,
                             type = 'salmon')


deseq_obj <- DESeq2::DESeqDataSetFromTximport(counts, colData = metadata, design = ~ type)

deseq 
```


Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

